rfcs/2020-05-25-2685-dev-workflow-simplification.md

```diff
@@ -0,0 +1,244 @@
+
+It proposes these changes such that our team and contributors will have to explicitly **opt in** to virtualization/containerization/shell-manipulation, and attempts to do it in a convienent, unsurprising way.
+
+## Motivation
+
+Currently, Vector's build system is a wild, powerful beast with a lot of amazing features and mechanisms. Unfortunately, complexity has kind of gotten away from us, and we need to reclaim ownership of the automations.
+
+Notably, our system works right now, as is, and we are not strongly motivated to replace or rebuild things just for the fun of it. We are motivated to make small, forward looking changes that improve the maintainability of our product.
+
+**Before we do this, we should make sure we agree this offers realistic benefits to our future maintenance.**
+
+## Guide-level Proposal
+
+> **Note:** Replaces the **Development → Setup** section of the `[CONTRIBUTING.md](http://contributing.md)` file.
+
+We're super excited to have you interested in working on Vector! Before you start you should pick how you want to develop.
+
+For small or first-time contributions, we recommend the Docker method. If you do a lot of contributing, try adopting the Nix method! It'll be way faster and feel more smooth. Prefer to do it yourself? That's fine too!
+
+### Use a Docker or Podman environment
+
+> **Targets:** You can use this method to produce AARCH64, Arm6/7, as well as x86/64 Linux builds.
+
+Since not everyone has a full working native environment, or can use Nix, we took our Nix environment and stuffed it into a Docker (or Podman) container!
+
+This is ideal for users who want it to "Just work" and just want to start contributing. It's also what we use for our CI, so you know if it breaks we can't do anything else until we fix it. 😉
+
+**Before you go farther, install Docker or Podman through your official package manager, or from the [Docker](https://docs.docker.com/get-docker/) or [Podman](https://podman.io/) sites.**
+
+```bash
+# Optional: Only if you use `podman`
+export container-tool="podman"
+```
+
+By default, `make environment` style tasks will do a `docker pull` from Github's container repository, you can **optionally** build your own environment while you make your morning coffee ☕:
+
+```bash
+# Optional: Only if you want to go make a coffee
+make environment-prepare
+```
+
+Now that you have your coffee, you can run any of the following:
+
+```bash
+# Enter a shell with optimized mounts for interactive processes.
+# Inside here, you can use Vector like you have full toolchain (See below!)
+make environment
+# Add extra cli opts
+make enviroment cli-opts="--publish 3000:2000"
+# Validate your code can compile
+make enviroment-check
```

Comment:
So, we'll have both `make enviroment-check` and `make check`? Can't we further reduce the duplication and entry points?

I don't know if you considered this option already, but can we decouple the environment and commands completely, and handle it via composition at the user level:

```shell
host$ make environment # drops the into the environment shell, possibly in a container
envshell$ make check # performs the check inside the environment
```

Reply:
Right, we have `make environment-check`, to run in the env, and `make check` to run on the host.

What you wrote works exactly fine, `make environment-check` just does the same thing as a shorthand since it's mean to be called from the host directly. Primarily, I suspect we'll all use `make check`, but all also use `make generate`.

Ideally, you'd install Nix and just use that though. :)

Reply:
Hmm, got it. It's quite obvious, but did you think about doing something like `make environment check` instead of `make environment check`? I'd try to avoid having multiple definitions for `check` and `environment-check`. Or do you have something else in mind to eliminate that duplication?

Reply:
I would love to do `make environment check`!!! Sadly I'm not sure how'd that work with how make does tasks?

I also considered keeping the current ENV var method, but I think the explicit, per command opt in is useful so you don't accidently start container builds etc.

Reply:
I'm oversimplifying, but `environment` target could just set `RUN` to `scripts/environment.sh`. And if we have `RUN`  empty by default, and keep using is as prefix - we'd, I think, get what we want.

Reply:
Hmmm, let me play with the idea of overwriting the env variable....

Reply:
> Then `make environment blah` will first switch `RUN`, then invoke `$RUN blah` aka `scripts/environment.sh blah`.

Will it do that even in the presence of parallel make options (like having `MAKEFLAGS=-j8`)? That looks like a race condition. Better would be `make MODE=environment blah`.

Reply:
So I played with this a bit: 
![image](https://user-images.githubusercontent.com/130903/83186989-4346fc00-a0e2-11ea-9759-c8548bd1c51a.png)

This works, but we'd need to add some complexity around `build` etc to make sure artifacts are correctly exported.

Reply:
> we'd need to add some complexity around `build` etc to make sure artifacts are correctly exported.

:frowning:

What if, instead of adding magic to `build` we just add a separate target, like `environment-copy-artifacts` or sth?

IMO, having extra complexity in tasks might be the root cause of the problems that we have currently - a few iterations and we get the same exploding amounts combinations.

Reply:
Sure but then we're running `make environment generate environment-copy-artifacts` instead of `make environment-generate` and that's not awesome.

Reply:
This is why I was falling back to having `make environment-generate` etc. It's simple, explicit, opt in.

Reply:
Maybe `make generate ENVIRONMENT=true` is best.

Reply:
> Maybe `make generate ENVIRONMENT=true` is best.

I like this idea.

Reply:
> Sure but then we're running `make environment generate environment-copy-artifacts` instead of `make environment-generate` and that's not awesome.

The good thing about it is that disabling unnecessary steps becomes a very natural thing - just don't specify them. But then, there's no way to specify those commands to run by default... Or is there? Yeah, I'm pretty sure we could make a thing where we just do `make environment qwe`, and it runs `qwe-environment-post` step automatically if we're in the `environment` mode, and doesn't if we're at "host" mode.
Not insisting on this though, so if you like `make environment-qwe` better - let's discuss that.

If we go with `make environment-*`, I'd think we ideally want a wildcard definition for them, so that they all have the same predictable behavior, and, most importantly - that it's defined at one place, and not for every command. Not sure about how to achieve it with `make`, but at least from the perspective of the goals - I think this is what we aim for. Otherwise, every command will have two definitions again, that can potentially diverge.
Do you share this point of view?

Reply:
`make generate ENVIRONMENT=true` also works fine, I guess.